### PR TITLE
ONNX Export Support for MSELoss

### PR DESCRIPTION
**Pull request by KsenijaS**
Add ONNX export support for torch.nn.MSELoss.

### CODE REVIEW

File: torch/onnx/symbolic_opset12.py [hunk 1]
```diff
@@ -13,3 +13,7 @@
 def einsum(g, equation, tensor_list):
     tensors = sym_help._unpack_list(tensor_list)
     return g.op("Einsum", *tensors, equation_s=equation)
+
+@parse_args('s', 'v', 'v')
+def mse(g, reduction, input, target):
```
Comment (lara-hdr) [inline torch/onnx/symbolic_opset12.py]: the function name should be mse_loss not mse..
I tried running the tests that you added and saw that there was the op 'aten::broadcast_tensors' in the jit trace but we don't have export support for it.

Comment (lara-hdr) [reply torch/onnx/symbolic_opset12.py]: Even if the tests are disable, run them locally and verify that the graph exports correctly

File: torch/onnx/symbolic_opset12.py [hunk 2]
```diff
@@ -13,3 +13,7 @@
...
+    return g.op("MeanSquaredDistance", input, target, reduction=reduction)
```
Comment (lara-hdr) [inline torch/onnx/symbolic_opset12.py]: reduction is an int in aten::mse_loss but a string in ONNX. 
So we need to add some logic here to convert the int value to the corresponding string ("none", "mean", "sum")

Comment (lara-hdr) [inline torch/onnx/symbolic_opset12.py]: reduction is a string attribute, so make sure to specify g.op(... reduction_s=...) and not g.op(... reduction=...)

File: torch/onnx/symbolic_opset12.py [hunk 3]
```diff
@@ -13,3 +12,18 @@
 def einsum(g, equation, tensor_list):
     tensors = sym_help._unpack_list(tensor_list)
     return g.op("Einsum", *tensors, equation_s=equation)
+
+def broadcast_tensors(g, tensor_list):
+    tensors = sym_help._unpack_list(tensor_list)
+    if (tensors[0].type().dim() != 2 and tensors[0].type().sizes() != tensors[1].type().sizes()):
+        return _unimplemented("broadcast_tensors", "broadcast for tensors is only implemented in case when there are 2 tensors of the same size")
+    return g.op("prim::ListConstruct", *tensors)
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset12.py]: why is this returning a `prim::ListConstruct` instead of an ONNX op? Is this then handled somewhere else?
Also, is this supposed to be doing some broadcasting?

Comment (KsenijaS) [reply torch/onnx/symbolic_opset12.py]: It should probably just return tensor_list.
Actually it's not doing broadcasting, since mse loss implementation in PyTorch requires tensors to be broadcastable, this op just check if the inputs are of the same size and then returns them, otherwise it throws an exception. This is only done in order to support mseloss export to ONNX and is not a complete implementation of broadcastable_tensors op

File: torch/onnx/symbolic_opset12.py [hunk 4]
```diff
@@ -21,6 +20,7 @@ def nll_loss(g, self, target, weight, reduction, ignore_index):
     reduction = sym_help._maybe_get_const(reduction, 'i')
     reduction_vals = ['none', 'mean', 'sum']
     reduction = reduction_vals[reduction]
+    return g.op("MeanSquaredDistance", input, target, reduction_s=reduction)
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset12.py]: nit: remove

File: torch/onnx/symbolic_opset12.py [hunk 5]
```diff
@@ -71,3 +71,18 @@ def nll_loss(g, self, target, weight, reduction, ignore_index):
 
 def nll_loss2d(g, self, target, weight, reduction, ignore_index):
     return nll_loss(g, self, target, weight, reduction, ignore_index)
+
+def broadcast_tensors(g, tensor_list):
+    tensors = sym_help._unpack_list(tensor_list)
+    if (tensors[0].type().dim() != 2 and tensors[0].type().sizes() != tensors[1].type().sizes()):
+        return _unimplemented("broadcast_tensors", "broadcast for tensors is only implemented in case when there are 2 tensors of the same size")
```
Comment (lara-hdr) [inline torch/onnx/symbolic_opset12.py]: _unimplemented() will throws an error like : "ONNX export failed on broadcast_tensors because broadcast for tensors is only implemented in case when there are 2 tensors of the same size not supported" which is not grammatically correct.
https://github.com/pytorch/pytorch/blob/afa8cbf8c21a080f43acfdf3cac6a07a4c606841/torch/onnx/symbolic_helper.py#L175

File: torch/onnx/symbolic_opset12.py (hunk 3 above)
Comment (KsenijaS) [reply torch/onnx/symbolic_opset12.py]: mean square distance accepts only 2 tensors.

File: torch/onnx/symbolic_opset12.py (hunk 4 above)
Comment (KsenijaS) [reply torch/onnx/symbolic_opset12.py]: remove what?

Comment (lara-hdr) [reply torch/onnx/symbolic_opset12.py]: this is the nll_loss function, line 23 returns MeanSquaredDistance

File: torch/onnx/symbolic_opset12.py [hunk 6]
```diff
@@ -71,3 +70,18 @@ def nll_loss(g, self, target, weight, reduction, ignore_index):
 
 def nll_loss2d(g, self, target, weight, reduction, ignore_index):
     return nll_loss(g, self, target, weight, reduction, ignore_index)
+
+def broadcast_tensors(g, tensor_list):
```
Comment (lara-hdr) [inline torch/onnx/symbolic_opset12.py]: could you leave a comment explaining what we are doing here and why we have a partial support of this op

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 7]
```diff
@@ -2642,6 +2642,36 @@ def forward(self, *tensor_list):
         x = torch.randn(3, 4)
         self.run_test(EinsumModelTranspose(), input=(x,))
 
+    @unittest.skip("Enable this once ORT version is updated")
```
Comment (lara-hdr) [inline test/onnx/test_pytorch_onnx_onnxruntime.py]: could you add a test in test_operators.py as well to make sure that broadcast_tensor is being exported correctly

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 8]
```diff
@@ -2642,6 +2642,36 @@ def forward(self, *tensor_list):
...
+        self.run_test(MSELossSum(), input=(x, y))
+
+        class MSELossMean(torch.nn.Module):
+            def forward(self, input, target):
+                loss = torch.nn.MSELoss()
+                return loss(input, target)
+
+        x = torch.randn(3, 5, requires_grad=True)
+        y = torch.randn(3, 5, requires_grad=True)
+        self.run_test(MSELossMean(), input=(x, y))
```
Comment (lara-hdr) [inline test/onnx/test_pytorch_onnx_onnxruntime.py]: add a test case with requires_grad=False for the target? (and add a small comment explaining that the output graph is different if  requires_grad=True/False for the target)

**Comment by lara-hdr** [comment]
Looks good. I have some minor comments inline.

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 8 above)
Comment (lara-hdr) [reply test/onnx/test_pytorch_onnx_onnxruntime.py]: I meant add a test with requires_grad=False in test_pytorch_onnx_onnxruntime. The current test all have target.requires_grad=True so will not test the mse op

File: torch/onnx/symbolic_opset12.py [hunk 9]
```diff
@@ -71,3 +70,21 @@ def nll_loss(g, self, target, weight, reduction, ignore_index):
 
 def nll_loss2d(g, self, target, weight, reduction, ignore_index):
     return nll_loss(g, self, target, weight, reduction, ignore_index)
+
+# This implementation of broadcast_tensors only covers the case when there are only 2 tensors
+# of the same size. This is needed for the export of mse_loss function and is only a partial
+# implementation of broadcast_tensors.
+def broadcast_tensors(g, tensor_list):
+    tensors = sym_help._unpack_list(tensor_list)
+    if (tensors[0].type().dim() != 2 and tensors[0].type().sizes() != tensors[1].type().sizes()):
```
Comment (lara-hdr) [inline torch/onnx/symbolic_opset12.py]: shouldn't this be an "or"?

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 8 above)
Comment (KsenijaS) [reply test/onnx/test_pytorch_onnx_onnxruntime.py]: there is already test when requires_grad=False

**Review by houseroad — COMMENTED** [review]
Test is still failing.

**Comment by KsenijaS** [comment]
> Test is still failing.

My SoftmaxCrossEntropy PR is reverted, and that PR updates onnx module, so probably that's why test is failing

**Comment by KsenijaS** [comment]
Closing this PR since MSELoss operator won't be included into ONNX 1.7
